common/plugin_control.c

```diff
@@ -68,6 +70,142 @@ const int64_t mistral_max_size = SSIZE_MAX;
  */
 struct timeval mistral_plugin_end;
 
+/* The actual log levels for each module. Initialized to
+ * DR_LOG_LEVEL_MAX so that before we have obtained the actual log level
+ * settings (from ELLEXUS_LOG_LEVEL) we capture every log message in the
+ * queue.
+ */
+
+#define LOG_LEVEL_INIT(module) DR_LOG_LEVEL_MAX,
+
+int log_level[LOG_module_MAX] = {
+    LOG_MODULES(LOG_LEVEL_INIT)
+};
+
+/* To set per-module logging levels we need to have in hand the module
+ * names, and level names, and for efficiency their lengths. */
+
+static struct {
+    const char *name;
+    size_t len;
+} log_module_names[] = {
+    #define MODULE_NAME(module) { #module, sizeof(#module)-1 },
+    LOG_MODULES(MODULE_NAME)
+    {
+        NULL, 0
+    }
+};
+
+static struct {
+    const char *name;
+    size_t len;
+} log_level_names[] = {
+    #define LEVEL_NAME(level) { #level, sizeof(#level)-1 },
+    LOG_LEVELS(LEVEL_NAME)
+    {
+        NULL, 0
+    }
+};
+
+/* Store the lengths of the scopes to avoid using strlen */
+size_t mistral_call_type_len[CALL_TYPE_MAX] = {
+    #define X(name, str) sizeof(str) - 1,
+    CALL_TYPE(X)
+    #undef X
+};
+/* Create various string arrays based off of the mistral_plugin.h header */
+const char * const mistral_contract_name[] = {
+    #define X(name, str, header) str,
+    CONTRACT(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_contract_header[] = {
+    #define X(name, str, header) header,
+    CONTRACT(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_scope_name[] = {
+    #define X(name, str) str,
+    SCOPE(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_measurement_name[] = {
+    #define X(name, str) str,
+    MEASUREMENT(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_unit_class_name[] = {
+    #define X(name, str) str,
+    UNIT_CLASS(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_unit_suffix[] = {
+    #define X(name, str, scale, type) str,
+    UNIT(X)
+    #undef X
+    NULL
+};
+
+const char * const mistral_call_type_name[] = {
+    #define X(name, str) str,
+    CALL_TYPE(X)
+    #undef X
+    NULL
+};
+
+const uint32_t mistral_call_type_mask[] = {
+    #define X(name, str) BITMASK(CALL_TYPE_ ## name),
+    CALL_TYPE(X)
+    #undef X
+    BITMASK(CALL_TYPE_MAX)
+};
+
+char mistral_call_type_names[CALL_TYPE_MASK_MAX][
+    #define X(name, str) + sizeof(str) + 1
+    CALL_TYPE(X)
+#undef X
+] = {
+    #define X(name, str) "",
+    CALL_TYPE(X)
+    #undef X
+};
+
+/* Similarly create some constant integer arrays */
+const uint32_t mistral_unit_scale[] = {
```

Comment:
I have all sorts of gripes about these arrays and how they are created, but this PR is not the right place for them.